File: android/app/src/main/kotlin/com/example/news_app/native/device/DeviceNativeHandler.kt
package com.nytimes.news_app.native.device

import android.content.Context
import com.nytimes.news_app.native.BaseNativeHandler
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.EventChannel
import kotlinx.coroutines.*

class DeviceNativeHandler(
    flutterEngine: FlutterEngine,
    context: Context
) : BaseNativeHandler(flutterEngine, context) {

    override val channelName: String = "device_service"
    override val eventChannels: Map<String, String> = mapOf()

    private lateinit var provider: DeviceInfoProvider

    override fun initialize() {
        super.initialize()
        provider = DeviceInfoProvider(context)
    }

    override fun handleMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getDeviceInfo" -> {
                try {
                    // Async method with location
                    provider.getDeviceInfo { info ->
                        GlobalScope.launch(Dispatchers.Main) {
                            result.success(info)
                        }
                    }
                } catch (e: Exception) {
                    result.error("GET_DEVICE_INFO_ERROR", "Failed to get device info: ${e.message}", null)
                }
            }

            "getDeviceInfoSync" -> {
                try {
                    // Sync method without full location
                    val info = provider.getDeviceInfoSync()
                    result.success(info)
                } catch (e: Exception) {
                    result.error("GET_DEVICE_INFO_SYNC_ERROR", "Failed to get device info sync: ${e.message}", null)
                }
            }

            "getSystemInfo" -> {
                try {
                    val info = provider.getSystemInfo()
                    result.success(info)
                } catch (e: Exception) {
                    result.error("GET_SYSTEM_INFO_ERROR", "Failed to get system info: ${e.message}", null)
                }
            }

            "getDeviceIdentifier" -> {
                try {
                    val identifier = provider.getDeviceIdentifier()
                    result.success(identifier)
                } catch (e: Exception) {
                    result.error("GET_DEVICE_ID_ERROR", "Failed to get device identifier: ${e.message}", null)
                }
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    override fun getStreamHandler(eventName: String): EventChannel.StreamHandler? {
        return null // Không cần stream handlers cho bây giờ
    }
}

File: ios/Runner/Native/Device/DeviceNativeHandler.swift
// ios/Runner/Native/Device/DeviceNativeHandler.swift
import Flutter
import UIKit

class DeviceNativeHandler: AbstractNativeHandler {
    override var channelName: String {
        return "device_service"
    }
    
    override var eventChannels: [String: String] {
        return [:] // Không cần event channels cho bây giờ
    }
    
    private let provider = DeviceInfoProvider()
    
    override func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getDeviceInfo":
            // Async version with location
            provider.getDeviceInfo { deviceInfo in
                DispatchQueue.main.async {
                    result(deviceInfo)
                }
            }
            
        case "getDeviceInfoSync":
            // Sync version without full location (fallback)
            let info = provider.getDeviceInfoSync()
            result(info)
            
        case "getSystemInfo":
            let info = provider.getSystemInfo()
            result(info)
            
        case "getDeviceIdentifier":
            let identifier = provider.getDeviceIdentifier()
            result(identifier)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    override func getStreamHandler(eventName: String) -> (any FlutterStreamHandler & NSObjectProtocol)? {
        return nil // Không cần stream handlers cho bây giờ
    }
}
